sys/stat_test.go

```diff
@@ -0,0 +1,156 @@
+	testFS := fstest.MapFS{
+		"dir": {
+			Mode:    osDirInfo.Mode(),
+			ModTime: osDirInfo.ModTime(),
+		},
+		"dir/file": {
+			Data:    fileData,
+			Mode:    osFileInfo.Mode(),
+			ModTime: osFileInfo.ModTime(),
+		},
+		"dir/file-sys": {
+			// intentionally skip other fields to prove sys is read.
+			Sys: &osFileSt,
+		},
+	}
+
+	fsDirInfo, err := testFS.Stat("dir")
+	require.NoError(t, err)
+	fsFileInfo, err := testFS.Stat("dir/file")
+	require.NoError(t, err)
+	fsFileInfoWithSys, err := testFS.Stat("dir/file-sys")
+	require.NoError(t, err)
+
+	tests := []struct {
+		name            string
+		info            fs.FileInfo
+		expectDevIno    bool
+		expectedMode    fs.FileMode
+		expectedSize    int64
+		expectAtimCtime bool
+	}{
+		{
+			name:            "os dir",
+			info:            osDirInfo,
+			expectDevIno:    true,
+			expectedMode:    fs.ModeDir | 0o0700,
+			expectedSize:    osDirInfo.Size(), // OS dependent
+			expectAtimCtime: true,
+		},
+		{
+			name:            "fs dir",
+			info:            fsDirInfo,
+			expectDevIno:    false,
+			expectedMode:    fs.ModeDir | 0o0700,
+			expectedSize:    0,
+			expectAtimCtime: false,
+		},
+		{
+			name:            "os file",
+			info:            osFileInfo,
+			expectDevIno:    true,
+			expectedMode:    0o0400,
+			expectedSize:    int64(len(fileData)),
+			expectAtimCtime: true,
+		},
+		{
+			name:            "fs file",
+			info:            fsFileInfo,
+			expectDevIno:    false,
+			expectedMode:    0o0400,
+			expectedSize:    int64(len(fileData)),
+			expectAtimCtime: false,
+		},
+		{
+			name:            "fs file with Stat_t in Sys",
+			info:            fsFileInfoWithSys,
+			expectDevIno:    true,
+			expectedMode:    0o0400,
+			expectedSize:    int64(len(fileData)),
+			expectAtimCtime: true,
+		},
+		{
+			name:            "os symlink",
+			info:            osSymlinkInfo,
+			expectDevIno:    true,
+			expectedMode:    fs.ModeSymlink,
+			expectedSize:    int64(len(file)),
+			expectAtimCtime: true,
+		},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+		t.Run(tc.name, func(t *testing.T) {
+			st := NewStat_t(tc.info)
+			if tc.expectDevIno && runtime.GOOS != "windows" {
+				require.NotEqual(t, uint64(0), st.Dev)
+				require.NotEqual(t, uint64(0), st.Ino)
+			} else {
+				require.Zero(t, st.Dev)
+				require.Zero(t, st.Ino)
+			}
+
+			// link mode may differ on windows, so mask
+			require.Equal(t, tc.expectedMode, st.Mode&tc.expectedMode)
+
+			if sysParseable && runtime.GOOS != "windows" {
+				switch st.Nlink {
+				case 2, 4: // dirents may include dot entries.
+					require.Equal(t, fs.ModeDir, st.Mode.Type())
+				default:
+					require.Equal(t, uint64(1), st.Nlink)
+				}
+			} else { // Nlink is possibly wrong, but not zero.
+				require.Equal(t, uint64(1), st.Nlink)
+			}
+
+			require.Equal(t, tc.expectedSize, st.Size)
+
+			if tc.expectAtimCtime && sysParseable {
+				// We don't validate times strictly because it is os-dependent
```

Comment:
I tried hard to do some validation on times, but it was a fools errand. it is really hard to get these to change especially without sleeping, in a portable way.